File: lib/python/src/aria/ops/suite_api_client.py
#  Copyright 2022 VMware, Inc.
#  SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import json
import logging
import math
from types import TracebackType
from typing import Any
from typing import Callable
from typing import Dict
from typing import Optional
from typing import Type

import requests
import urllib3
from aria.ops.object import Identifier
from aria.ops.object import Key
from aria.ops.object import Object
from requests import Response

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

logger = logging.getLogger(__name__)


class SuiteApiConnectionParameters:
    def __init__(
        self, host: str, username: str, password: str, auth_source: str = "LOCAL"
    ):
        """Initialize SuiteApi Connection Parameters

        Args:
            host (str): The host to use for connecting to the SuiteAPI.
            username (str): Username used to authenticate to SuiteAPI
            password (str): Password used to authenticate to SuiteAPI
            auth_source (str): Source of authentication. Defaults to "LOCAL"
        """
        if "http" in host:
            self.host = f"{host}/suite-api/"
        else:
            self.host = f"https://{host}/suite-api/"
        self.username = username
        self.password = password
        self.auth_source = auth_source


class SuiteApiClient:
    """Class for simplifying calls to the SuiteAPI

    Automatically handles:
    * Token based authentication
    * Required headers
    * Releasing tokens (when used in a 'with' statement)
    * Paging (when using 'paged_get' or 'paged_post')
    * Logging requests

    This class is intended to be used in a with statement:
    with VROpsSuiteAPIClient() as suiteApiClient:
        # Code using suiteApiClient goes here
        ...
    """

    def __init__(self, connection_params: SuiteApiConnectionParameters):
        """Initializes a SuiteAPI client.

        Args:
             connection_params (SuiteApiConnectionParameters): Connection parameters for the Suite API.
        """
        self.credential = connection_params
        self.token = ""

    def __enter__(self) -> SuiteApiClient:
        """Acquire a token upon entering the 'with' context

        Returns:
            SuiteApiClient: The current instance of the class.
        """
        self.token = self.get_token()
        return self

    def __exit__(
        self,
        exception_type: Optional[Type[BaseException]],
        exception_value: Optional[BaseException],
        traceback: Optional[TracebackType],
    ) -> None:
        """Release the token upon exiting the 'with' context

        Args:
            exception_type (Optional[Type[BaseException]]): Unused
            exception_value (Optional[BaseException]): Unused
            traceback (Optional[TracebackType]): Unused
        """
        self.release_token()

    def get_token(self) -> str:
        """Get the authentication token

        Gets the current authentication token. If no current token exists, acquires an authentication token first.

        Returns:
             The authentication token
        """
        if self.token == "":
            with self.post(
                "/api/auth/token/acquire",
                json={
                    "username": self.credential.username,
                    "password": self.credential.password,
                    "authSource": self.credential.auth_source,
                },
            ) as token_response:
                if token_response.ok:
                    self.token = token_response.json()["token"]
                    logger.debug("Acquired token " + self.token)
                else:
                    logger.warning(
                        f"Could not acquire SuiteAPI token: {token_response}"
                    )

        return self.token

    def release_token(self) -> None:
        """Release the authentication token, if it exists"""

        if self.token != "":
            self.post("auth/token/release").close()
            self.token = ""

    def get(self, url: str, **kwargs: Any) -> Response:
        """Send a GET request to the SuiteAPI
        The 'Response' object should be used in a 'with' block or
        manually closed after use

        Args:
            url (str): URL to send GET request to
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
            The API response
        """
        return self._request_wrapper(requests.get, url, **kwargs)

    def paged_get(self, url: str, key: str, **kwargs: Any) -> dict:
        """Send a GET request to the SuiteAPI that gets a paged response

        Args:
            url (str): URL to send GET request to
            key (str): Json key that contains the paged data
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        return self._paged_request(requests.get, url, key, **kwargs)

    def post(self, url: str, **kwargs: Any) -> Response:
        """Send a POST request to the SuiteAPI
        The 'Response' object should be used in a 'with' block or
        manually closed after use

        Args:
            url (str): URL to send POST request to
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        kwargs.setdefault("headers", {})
        kwargs["headers"].setdefault("Content-Type", "application/json")
        return self._request_wrapper(requests.post, url, **kwargs)

    def paged_post(self, url: str, key: str, **kwargs: Any) -> dict:
        """Send a POST request to the SuiteAPI that gets a paged response.

        Args:
            url (str): URL to send POST request to
            key (str): Json key that contains the paged data
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        kwargs.setdefault("headers", {})
        kwargs["headers"].setdefault("Content-Type", "application/json")
        return self._paged_request(requests.post, url, key, **kwargs)

    def put(self, url: str, **kwargs: Any) -> Response:
        """Send a PUT request to the SuiteAPI
        The 'Response' object should be used in a 'with' block or
        manually closed after use

        Args:
            url (str): URL to send PUT request to
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        return self._request_wrapper(requests.put, url, **kwargs)

    def patch(self, url: str, **kwargs: Any) -> Response:
        """Send a PATCH request to the SuiteAPI
        The 'Response' object should be used in a 'with' block or
        manually closed after use

        Args:
            url (str): URL to send PATCH request to
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
            The API response
        """
        return self._request_wrapper(requests.patch, url, **kwargs)

    def delete(self, url: str, **kwargs: Any) -> Response:
        """Send a DELETE request to the SuiteAPI
        The 'Response' object should be used in a 'with' block or
        manually closed after use

        Args:
            url (str): URL to send DELETE request to
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        return self._request_wrapper(requests.delete, url, **kwargs)

    def _add_paging(self, **kwargs: Any) -> dict:
        kwargs.setdefault("params", {})
        kwargs["params"].setdefault("page", 0)
        kwargs["params"].setdefault("pageSize", 1000)

        if "page" in kwargs:
            kwargs["params"]["page"] = kwargs.pop("page")
        if "pageSize" in kwargs:
            kwargs["params"]["pageSize"] = kwargs.pop("pageSize")

        return kwargs

    # Implementations for common endpoints:

    def query_for_resources(self, query: Dict[str, Any]) -> list[Object]:
        """Query for resources using the Suite API, and convert the
        responses to SDK Objects.

        Note that not all information from the query is returned. For example, the
        query returns health statuses of each object, but those are not present in
        the resulting Objects. If information other than the Object itself is needed,
        you will need to call the endpoint and process the results manually.

        Args:
            query (Dict[str, Any]): json of the resourceQuery, as defined in the SuiteAPI docs:
                https://[[aria-ops-hostname]]/suite-api/doc/swagger-ui.html#/Resources/getMatchingResourcesUsingPOST

        Returns:
             list of sdk Objects representing each of the returned objects.
        """
        try:
            results = []
            if "name" in query and "regex" in query:
                # This is behavior in the suite api itself, we're just warning about it
                # here to avoid confusion.
                logger.warning(
                    "'name' and 'regex' are mutually exclusive in resource "
                    "queries. Ignoring the 'regex' key in favor of 'name' "
                    "key."
                )
            # The 'name' key takes an array but only looks up the first element.
            # Fix that limitation here.
            if "name" in query and len(query["name"]) > 1:
                json_body = query.copy()
                # TODO: Improve concurrancy when we add async support
                #  to suite_api_client
                for name in query["name"]:
                    json_body.update({"name": [name]})
                    response = self.paged_post(
                        "/api/resources/query", "resourceList", json=json_body
                    )
                    results.extend(response.get("resourceList", []))
            else:
                response = self.paged_post(
                    "/api/resources/query",
                    "resourceList",
                    json=query,
                )
                results = response.get("resourceList", [])
            return [key_to_object(obj["resourceKey"]) for obj in results]
        except Exception as e:
            logger.error(e)
            logger.exception(e)
            return []

    def _paged_request(
        self, request_func: Callable, url: str, key: str, **kwargs: Any
    ) -> dict:
        """Send a request to the SuiteAPI that returns a paged response. Each response must have data returned in an
        array at key 'key'. The array from the responses will be combined into a single array and returned in a map of
        the form:
        {
           "{key}": [aggregated data]
        }

        Args:
            url(str): URL to send request to
            key (str): Json key that contains the paged data
            kwargs (Any): Additional keyword arguments to pass to request

        Returns:
             The API response
        """
        kwargs = self._add_paging(**kwargs)
        with self._request_wrapper(request_func, url, **kwargs) as page_0:
            if page_0.status_code < 300:
                page_0_body = json.loads(page_0.text)
            else:
                # _request_wrapper will log the error
                # TODO: How should we communicate to caller that
                #       request(s) have failed?
                return {key: []}
        total_objects = int(
            page_0_body.get("pageInfo", {"totalCount": 1}).get("totalCount", 1)
        )
        page_size = kwargs["params"]["pageSize"]
        remaining_pages = math.ceil(total_objects / page_size) - 1
        objects = page_0_body.get(key, [])
        while remaining_pages > 0:
            kwargs = self._add_paging(page=remaining_pages, **kwargs)
            with self._request_wrapper(request_func, url, **kwargs) as page_n:
                if page_n.status_code < 300:
                    page_n_body = json.loads(page_n.text)
            objects.extend(page_n_body.get(key, []))
            remaining_pages -= 1
        return {key: objects}

    def _request_wrapper(
        self, request_func: Callable[..., Response], url: str, **kwargs: Any
    ) -> Response:
        kwargs = self._to_vrops_request(url, **kwargs)
        result = request_func(**kwargs)
        if result.ok:
            logger.info(
                f"{request_func.__name__} {kwargs['url']}: OK({result.status_code})"
            )
        else:
            logger.warning(
                f"{request_func.__name__} {kwargs['url']}: ERROR({result.status_code})"
            )
        logger.debug(result.text)
        return result

    def _to_vrops_request(self, url: str, **kwargs: Any) -> dict:
        kwargs.setdefault("url", url)
        kwargs.setdefault("headers", {})
        if self.token:
            kwargs["headers"]["Authorization"] = "vRealizeOpsToken " + self.token
        kwargs["headers"].setdefault("Accept", "application/json")
        kwargs.setdefault("verify", False)

        url = kwargs["url"]
        if "internal/" in url:
            kwargs["headers"]["X-vRealizeOps-API-use-unsupported"] = "true"
            logger.info(f"Using unsupported API: {url}")
        if url.startswith("http"):
            return kwargs

        if url.startswith("/"):
            url = url[1:]
        if url.startswith("suite-api/"):
            url = url[10:]
        elif url.startswith("api") or url.startswith("internal"):
            kwargs["url"] = self.credential.host + url
        else:
            kwargs["url"] = self.credential.host + "api/" + url
        return kwargs


# Helper methods:


def key_to_object(json_object_key: Dict[str, Any]) -> Object:
    return Object(
        Key(
            json_object_key["adapterKindKey"],
            json_object_key["resourceKindKey"],
            json_object_key["name"],
            [
                Identifier(
                    identifier["identifierType"]["name"],
                    identifier["value"],
                    identifier["identifierType"]["isPartOfUniqueness"],
                )
                for identifier in json_object_key["resourceIdentifiers"]
            ],
        )
    )

File: lib/java/AdapterLibrary/src/main/kotlin/com/vmware/aria/operations/SuiteApiClient.kt
/*
 * Copyright 2023 VMware, Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.vmware.aria.operations

import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.cio.CIO
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.accept
import io.ktor.client.request.delete
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.apache.logging.log4j.Logger
import java.security.cert.X509Certificate
import javax.net.ssl.X509TrustManager
import kotlin.math.ceil
import kotlin.math.roundToInt
import kotlin.reflect.KProperty

/**
 * Class to hold Suite API connection information
 * @property hostname The hostname or IP address of the SuiteAPI
 * @property username The username to connect with
 * @property password The username's password
 * @property port Determines the port to connect to the server with. Used for testing.
 * @property verify Determines if the SSL Certificate from the server should be verified. Used for testing.
 */
@Serializable
data class SuiteApiConnectionInfo @JvmOverloads constructor(
    @SerialName("host_name") val hostname: String,
    @SerialName("user_name") val username: String,
    val password: String,
    val port: Int = 443,
    val verify: Boolean = false,
) {
    @Transient
    private val baseUrl: String = getBaseUrl()

    @Transient
    internal val authenticationSource = "LOCAL"

    private fun getBaseUrl(): String {
        var url = hostname
        if (!url.startsWith("https://")) {
            url = "https://$url"
        }
        if (url.endsWith(URL_PATH_SEPARATOR)) {
            url = url.substringBefore(URL_PATH_SEPARATOR)
        }
        url = "$url:$port$URL_PATH_SEPARATOR"
        return url
    }

    /**
     * Given an API endpoint, prepend the protocol and hostname. If the endpoint does not
     * include 'suite-api/', that is added between the hostname and endpoint (unless the
     * endpoint is an internal API).
     * @param endpoint Endpoint to turn into a full URL
     * @return Full URL of the given endpoint
     */
    fun getUrl(endpoint: String): String {
        if (!endpoint.contains("suite-api") && !endpoint.contains("internal")) {
            "${baseUrl}suite-api$URL_PATH_SEPARATOR"
        } else {
            baseUrl
        }.let { baseUrl ->
            return if (baseUrl.endsWith(URL_PATH_SEPARATOR) &&
                endpoint.startsWith(URL_PATH_SEPARATOR)
            ) {
                "${baseUrl.dropLast(1)}$endpoint"
            } else if (baseUrl.endsWith(URL_PATH_SEPARATOR) ||
                endpoint.startsWith(URL_PATH_SEPARATOR)
            ) {
                "$baseUrl$endpoint"
            } else {
                "$baseUrl${URL_PATH_SEPARATOR}$endpoint"
            }
        }
    }

    companion object {
        private const val URL_PATH_SEPARATOR = "/"
    }
}

/**
 * Gets a new SuiteAPI Authentication token if one does not exist, or if the current
 * token will expire within 10 seconds
 * @property logger A logger
 * @property client An HTTP Client used for requesting and deleting tokens
 * @property connectionInfo A [SuiteApiConnectionInfo] class containing the Suite API host and credentials
 */
private class TokenDelegate(
    private val logger: Logger,
    private val client: HttpClient,
    private val connectionInfo: SuiteApiConnectionInfo,
) {
    /**
     * Class representing the Json body of the post request for acquiring a token
     */
    @Serializable
    private data class RequestBody(
        val username: String,
        val password: String,
        val authSource: String,
    )

    /**
     * Class representing the Json response of the post request for acquiring a token
     */
    @Serializable
    private data class TokenInfo(
        val validity: Long,
        val token: String,
        val expiresAt: String,
        val roles: List<String>,
    )

    private var tokenInfo: TokenInfo? = null
    private val requestBody = RequestBody(
        connectionInfo.username,
        connectionInfo.password,
        connectionInfo.authenticationSource
    )
    private val mutex = Mutex()

    /**
     * Returns true if a token exists and will not expire within the next 10 seconds
     */
    private fun isValid(): Boolean = tokenInfo?.let { token ->
        // Add 10 seconds so that the token won't expire between retrieval and use
        token.validity > System.currentTimeMillis() + 10_000L
    } ?: false

    /**
     * Returns the token if it exists, or requests one if it does not exist. If multiple
     * requests for the token are made concurrently, only the first will request a token,
     * and the other requests will wait until that request has completed.
     */
    operator fun getValue(thisRef: Any?, property: KProperty<*>): String =
        runBlocking {
            mutex.withLock {
                if (!isValid()) {
                    logger.debug("Requesting New Token")
                    tokenInfo =
                        client.post(connectionInfo.getUrl("api/auth/token/acquire")) {
                            accept(ContentType.Application.Json)
                            contentType(ContentType.Application.Json)
                            setBody(requestBody)
                        }.throwOnErrorStatus().body()
                }
                return@runBlocking tokenInfo?.token ?: ""
            }
        }

    /**
     * Releases a token, so that it is no longer valid.
     */
    fun releaseToken(): Unit = runBlocking {
        logger.debug("Releasing Token")
        if (isValid()) {
            client.post(connectionInfo.getUrl("/api/auth/token/release")) {
                contentType(ContentType.Application.Json)
                accept(ContentType.Application.Json)
                header(
                    "Authorization",
                    "vRealizeOpsToken ${tokenInfo?.token}"
                )
            }
        }
        tokenInfo = null
    }
}

/**
 * Class that simplifies making calls to VMware Aria Operations' Suite API
 * @property connectionInfo Connection and credential information for connecting to the SuiteAPI
 * @property maxConnections The maximum number of concurrent connections allowed. Defaults to 10
 */
class SuiteApiClient @JvmOverloads constructor(
    val connectionInfo: SuiteApiConnectionInfo,
    maxConnections: Int = 10,
) : AutoCloseable {
    private val client = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
            })
        }
        if (!connectionInfo.verify) {
            engine {
                https {
                    trustManager = object : X509TrustManager {
                        override fun checkClientTrusted(
                            chain: Array<out X509Certificate>?,
                            authType: String?,
                        ) = Unit

                        override fun checkServerTrusted(
                            chain: Array<out X509Certificate>?,
                            authType: String?,
                        ) = Unit

                        override fun getAcceptedIssuers(): Array<X509Certificate> =
                            emptyArray()

                    }
                }
            }
        }
    }
    private val tokenDelegate = TokenDelegate(getLogger(), client, connectionInfo)
    private val token by tokenDelegate
    private val logger = getLogger()
    private var throttler = Semaphore(maxConnections)

    /**
     * Sets the maximum number of concurrent connections allowed. The default value is 10.
     * Note: Any existing connections when this method is called will not be bound
     * by the new limit.
     * @param maxConnections The maximum number of concurrent connections allowed.
     */
    var maxConnections: Int = maxConnections
        set(maxConnections: Int) {
            if (field != maxConnections) {
                field = maxConnections
                this.throttler = Semaphore(field)
            }
        }

    /**
     * Performs a GET operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @return A [JsonObject] containing the content as a json object. Throws if the
     * response code is not 2xx.
     */
    @Throws(SuiteApiClientException::class)
    fun get(endpoint: String): JsonObject =
        runBlocking {
            getAsync(endpoint)
        }

    /**
     * A Kotlin [suspend] method that performs a GET operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @return A [JsonObject] containing the content as a json object. Throws if the
     * response code is not 2xx.
     */
    @Throws(SuiteApiClientException::class)
    suspend fun getAsync(endpoint: String): JsonObject {
        return getRawResponseAsync(endpoint).throwOnErrorStatus().toJson()
    }

    /**
     * Performs a GET operation on the given endpoint.
     * @param endpoint The API endpoint to perform a get request on
     */
    fun getRawResponse(endpoint: String): HttpResponse =
        runBlocking {
            getRawResponseAsync(endpoint)
        }

    /**
     * A Kotlin [suspend] method that performs a GET operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @return An [HttpResponse] containing the return code and content
     */
    suspend fun getRawResponseAsync(endpoint: String): HttpResponse {
        return throttler.throttleCall {
            client.get(connectionInfo.getUrl(endpoint)) {
                accept(ContentType.Application.Json)
                header("Authorization", "vRealizeOpsToken $token")
                if (endpoint.contains("internal")) {
                    logger.debug("Using unsupported API: $endpoint")
                    header("X-vRealizeOps-API-use-unsupported", true)
                }
            }
        }
    }

    /**
     * A method that performs a paged GET operation on the given endpoint.
     * After the first GET operation succeeds, subsequent GET requests will be done
     * concurrently (with a max of [maxConnections] concurrent requests) until all of
     * paged data has been retrieved.
     * @param endpoint The API endpoint to call
     * @param pagedArrayKey The key on each returned json object to combine. This key should
     * contain a json array with each element being the data that is being paged.
     * @param pageSize The number of elements to return with each page
     * @return A [JsonObject] containing the content as a json object. The object will
     * contain the key [pagedArrayKey] that contains all the paged data combined into
     * a single array.
     */
    @Throws(SuiteApiClientException::class)
    @JvmOverloads
    fun getPaged(
        endpoint: String,
        pagedArrayKey: String,
        pageSize: Int = 1000,
    ): JsonObject =
        runBlocking {
            getPagedAsync(endpoint, pagedArrayKey, pageSize)
        }

    /**
     * A Kotlin [suspend] method that performs a paged GET operation on the given endpoint.
     * After the first GET operation succeeds, subsequent GET requests will be done
     * concurrently (with a max of [maxConnections] concurrent requests) until all of
     * paged data has been retrieved.
     * @param endpoint The API endpoint to call
     * @param pagedArrayKey The key on each returned json object to combine. This key should
     * contain a json array with each element being the data that is being paged.
     * @param pageSize The number of elements to return with each page
     * @return A [JsonObject] containing the content as a json object. The object will
     * contain the key [pagedArrayKey] that contains all the paged data combined into
     * a single array.
     */
    @Throws(SuiteApiClientException::class)
    @JvmOverloads
    suspend fun getPagedAsync(
        endpoint: String,
        pagedArrayKey: String,
        pageSize: Int = 1000,
    ): JsonObject {
        val page0 = getAsync(endpoint.addPaging(0, pageSize))
        val total =
            page0["pageInfo"]?.jsonObject?.get("totalCount")?.jsonPrimitive?.int ?: 1
        val remainingPages =
            ceil(((total - pageSize.coerceAtMost(total)).toDouble() / pageSize.toDouble())).roundToInt()

        val elements =
            (listOf(page0) + (
                    (1..remainingPages).asyncMap("Get Paged $endpoint") { page ->
                        getAsync(endpoint.addPaging(page, pageSize))
                    }))
                .flatMap { it[pagedArrayKey]?.jsonArray ?: JsonArray(emptyList()) }

        return JsonObject(
            mapOf(
                "count" to JsonPrimitive(elements.size),
                pagedArrayKey to JsonArray(elements)
            )
        )
    }

    /**
     * A method that performs a POST operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @param jsonBody A json object containing the payload of the POST request.
     * @return A [JsonObject] containing the content as a json object. Throws if the
     * response code is not 2xx.
     */
    @Throws(SuiteApiClientException::class)
    fun post(endpoint: String, jsonBody: JsonObject): JsonObject =
        runBlocking {
            postAsync(endpoint, jsonBody)
        }

    /**
     * A Kotlin [suspend] method that performs a POST operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @param jsonBody A json object containing the payload of the POST request.
     * @return A [JsonObject] containing the content as a json object. Throws if the
     * response code is not 2xx.
     */
    @Throws(SuiteApiClientException::class)
    suspend fun postAsync(
        endpoint: String,
        jsonBody: JsonObject,
    ): JsonObject {
        return postRawResponseAsync(endpoint, jsonBody).throwOnErrorStatus().toJson()
    }

    /**
     * A method that performs a POST operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @param jsonBody A json object containing the payload of the POST request.
     * @return An [HttpResponse] containing the return code and content
     */
    fun postRawResponse(endpoint: String, jsonBody: JsonObject): HttpResponse =
        runBlocking {
            postRawResponseAsync(endpoint, jsonBody)
        }

    /**
     * A Kotlin [suspend] method that performs a POST operation on the given endpoint.
     * @param endpoint The API endpoint to call
     * @param jsonBody A json object containing the payload of the POST request.
     * @return An [HttpResponse] containing the return code and content
     */
    suspend fun postRawResponseAsync(
        endpoint: String,
        jsonBody: JsonObject,
    ): HttpResponse {
        return throttler.throttleCall {
            client.post(connectionInfo.getUrl(endpoint)) {
                accept(ContentType.Application.Json)
                header("Authorization", "vRealizeOpsToken $token")
                if (endpoint.contains("internal")) {
                    logger.debug("Using unsupported API: $endpoint")
                    header("X-vRealizeOps-API-use-unsupported", true)
                }
                contentType(ContentType.Application.Json)
                setBody(jsonBody)
            }
        }
    }

    /**
     * A method that performs a paged POST operation on the given endpoint.
     * After the first GET operation succeeds, subsequent GET requests will be done
     * concurrently (with a max of [maxConnections] concurrent requests) until all the
     * paged data has been retrieved.
     * @param endpoint The API endpoint to call
     * @param pagedArrayKey The key on each returned json object to combine. This key should
     * contain a json array with each element being the data that is being paged.
     * @param pageSize The number of elements to return with each page
     * @return A [JsonObject] containing the content as a json object. The object will
     * contain the key [pagedArrayKey] that contains all the paged data combined into
     * a single array.
     */
    @Throws(SuiteApiClientException::class)
    @JvmOverloads
    fun postPaged(
        endpoint: String,
        jsonBody: JsonObject,
        pagedArrayKey: String,
        pageSize: Int = 1000,
    ): JsonObject =
        runBlocking {
            postPagedAsync(endpoint, jsonBody, pagedArrayKey, pageSize)
        }

    /**
     * A Kotlin [suspend] method that performs a paged POST operation on the given endpoint.
     * After the first GET operation succeeds, subsequent GET requests will be done
     * concurrently (with a max of [maxConnections] concurrent requests) until all the
     * paged data has been retrieved.
     * @param endpoint The API endpoint to call
     * @param pagedArrayKey The key on each returned json object to combine. This key should
     * contain a json array with each element being the data that is being paged.
     * @param pageSize The number of elements to return with each page
     * @return A [JsonObject] containing the content as a json object. The object will
     * contain the key [pagedArrayKey] that contains all the paged data combined into
     * a single array.
     */
    @Throws(SuiteApiClientException::class)
    @JvmOverloads
    suspend fun postPagedAsync(
        endpoint: String,
        jsonBody: JsonObject,
        pagedArrayKey: String,
        pageSize: Int = 1000,
    ): JsonObject {
        val page0 = postAsync(endpoint.addPaging(0, pageSize), jsonBody)
        val total =
            page0["pageInfo"]?.jsonObject?.get("totalCount")?.jsonPrimitive?.int ?: 1
        val remainingPages =
            ceil(((total - pageSize.coerceAtMost(total)).toDouble() / pageSize.toDouble())).roundToInt()

        val elements =
            (listOf(page0) + (
                    (1..remainingPages).asyncMap("Post Paged $endpoint") { page ->
                        postAsync(
                            endpoint.addPaging(page, pageSize),
                            jsonBody
                        )
                    }))
                .flatMap { it[pagedArrayKey]?.jsonArray ?: JsonArray(emptyList()) }

        return JsonObject(
            mapOf(
                "count" to JsonPrimitive(elements.size),
                pagedArrayKey to JsonArray(elements)
            )
        )
    }

    /**
     * A method that performs a DELETE operation on the given endpoint.
     * @param endpoint The API endpoint to call
     */
    @Throws(SuiteApiClientException::class)
    fun delete(endpoint: String) =
        runBlocking {
            deleteAsync(endpoint)
        }

    /**
     * A Kotlin [suspend] method that performs a DELETE operation on the given endpoint.
     * @param endpoint The API endpoint to call
     */
    @Throws(SuiteApiClientException::class)
    suspend fun deleteAsync(endpoint: String) {
        return throttler.throttleCall {
            client.delete(connectionInfo.getUrl(endpoint)) {
                accept(ContentType.Application.Json)
                header("Authorization", "vRealizeOpsToken $token")
                if (endpoint.contains("internal")) {
                    logger.debug("Using unsupported API: $endpoint")
                    header("X-vRealizeOps-API-use-unsupported", true)
                }
            }.throwOnErrorStatus()
        }
    }

    /**
     * Query for resources using the Suite API, and convert the
     * responses to SDK Objects.
     *
     * Note that not all information from the query is returned. For example, the
     * query returns health statuses of each object, but those are not present in
     * the resulting Objects. If information other than the Object itself is needed,
     * you will need to call the endpoint and process the results manually.
     *
     * @param query: json of the resourceQuery, as defined in the SuiteAPI docs:
     * https://[[aria-ops-hostname]]/suite-api/doc/swagger-ui.html#/Resources/getMatchingResourcesUsingPOST
     * @return list of sdk [Objects][Object] representing each of the returned objects.
     */
    fun queryForResources(query: JsonObject): List<Object> {
        try {
            var results = mutableListOf<JsonElement>()
            if (query.containsKey("name") && query.containsKey("regex")) {
                // This is behavior in the suite api itself, we're just warning about it
                // here to avoid confusion.
                logger.warn(
                    "'name' and 'regex' are mutually exclusive in resource " +
                            "queries. Ignoring the 'regex' key in favor of 'name' " +
                            "key."
                )
            }
            // The 'name' key takes an array but only looks up the first element.
            // Fix that limitation here .
            if (query["name"]?.jsonArray?.size?.let { it > 1 } == true) {
                val jsonBody = query
                for (name in query["name"]?.jsonArray ?: emptyList()) {
                    val newJsonBody = jsonBody.toMutableMap()
                    newJsonBody["name"] = buildJsonArray { add(name) }
                    val response = postPaged(
                        "/api/resources/query",
                        Json.encodeToJsonElement(newJsonBody).jsonObject,
                        "resourceList"
                    )
                    results.addAll(
                        response["resourceList"]?.jsonArray ?: buildJsonArray { })
                }
            } else {
                val response = this.postPaged(
                    "/api/resources/query",
                    query,
                    "resourceList",
                )
                results.addAll(
                    response.jsonObject["resourceList"]?.jsonArray ?: buildJsonArray {})
            }
            return results.mapNotNull { obj ->
                obj.jsonObject["resourceKey"]?.jsonObject?.let {
                    keyToObject(it)
                }
            }
        } catch (e: Exception) {
            logger.error(e)
            return emptyList()
        }
    }

    private fun keyToObject(jsonObjectKey: JsonObject): Object? {
        return Object(
            Key(
                jsonObjectKey["adapterKindKey"]?.jsonPrimitive?.content ?: return null,
                jsonObjectKey["resourceKindKey"]?.jsonPrimitive?.content ?: return null,
                jsonObjectKey["name"]?.jsonPrimitive?.content ?: return null,
                jsonObjectKey["resourceIdentifiers"]?.jsonArray?.map { identifier: JsonElement ->
                    Identifier(
                        identifier.jsonObject["identifierType"]?.jsonObject?.get("name")?.jsonPrimitive?.content ?: return null,
                        identifier.jsonObject["value"]?.jsonPrimitive?.content ?: return null,
                        identifier.jsonObject["identifierType"]?.jsonObject?.get("isPartOfUniqueness")?.jsonPrimitive?.booleanOrNull ?: return null,
                    )
                } ?: emptyList()
            )
        )
    }


    /**
     * Releases the token if present and closes any features of the client that require closing
     */
    override fun close() {
        try {
            tokenDelegate.releaseToken()
        } catch (e: SuiteApiClientException) {
            logger.warn("Could not release token. This is likely because the token has already been invalidated. ${e.message}")
            logger.debug(
                "Could not release token. This is likely because the token has already been invalidated. ${e.message}",
                e
            )
        }
        client.close()
    }

    private fun String.addPaging(page: Int = 0, pageSize: Int = 1000): String =
        if (this.contains("?")) {
            "$this&page=$page&pageSize=$pageSize"
        } else {
            "$this?page=$page&pageSize=$pageSize"
        }

    private suspend fun <T> Semaphore.throttleCall(body: suspend () -> T): T {
        this.acquire()
        return try {
            body()
        } finally {
            this.release()
        }
    }
}

private suspend fun HttpResponse.toJson() =
    Json.parseToJsonElement(this.bodyAsText()).jsonObject

private suspend fun HttpResponse.throwOnErrorStatus(): HttpResponse {
    if (this.status.value >= 300) {
        throw SuiteApiClientException(
            "${this.call.request.method.value} request to " +
                    "${this.call.request.url} returned ${this.status.description} " +
                    "(${this.status.value})", this.status.value
        )
    }
    return this
}

private suspend inline fun <T, R> Iterable<T>.asyncMap(
    scopeName: String,
    crossinline transform: suspend CoroutineScope.(T) -> R,
): List<R> {
    val currentScope = CoroutineScope(currentCoroutineContext())
    val context = CoroutineName(scopeName)
    return map { currentScope.async(context) { transform(it) } }.awaitAll()
}
